runtime/runtime/src/congestion_control.rs

```diff
@@ -612,10 +593,49 @@ pub fn bootstrap_congestion_info(
     }))
 }
 
-impl DelayedReceiptQueueWrapper {
-    pub fn new(queue: DelayedReceiptQueue) -> Self {
+/// A wrapper around `DelayedReceiptQueue` to accumulate changes in gas and
+/// bytes.
+///
+/// This struct exists for two reasons. One, to encapsulate the accounting of
+/// gas and bytes in functions that can be called in all necessary places. Two,
+/// to accumulate changes and only apply them to `CongestionInfo` in the end,
+/// which avoids problems with multiple mutable borrows with the closure
+/// approach we currently have in receipt processing code.
+///
+/// We use positive added and removed values to avoid integer conversions with
+/// the associated additional overflow conditions.
+pub(crate) struct DelayedReceiptQueueWrapper<'a> {
+    // The delayed receipt queue.
+    queue: DelayedReceiptQueue,
+
+    // Protocol version, epoch_info_provider, shard_id, and epoch_id are used to
+    // determine if a receipt belongs to the current shard or not after a resharding event.
+    protocol_version: ProtocolVersion,
+    epoch_info_provider: &'a dyn EpochInfoProvider,
+    shard_id: ShardId,
+    epoch_id: EpochId,
```

Comment:
Could this be just `shard_layout` and `shard_id`?

Reply:
I guess the epoch info provider that is available in runtime doesn't expose the shard layout directly but instead just provides the `account_id_to_shard_id` method. I think it would be nice to refactor it to return the shard layout instead but that's perhaps out of scope of this PR. Additionally the provide could be strictly pegged to the current epoch and block from the chunk apply context instead of needing to provide the epoch id or prev_block_hash to every method. Let me know if this makes sense and I'll create an issue for this.

Reply:
Yes, that makes sense, I was just working with the epoch_info_provider that we currently had in runtime and didn't think about refactoring it.